### PR TITLE
client: add procession of SEEK_HOLE and SEEK_DATA in lseek.

### PR DESCRIPTION
**Pull request by shenhang**
Fixes: https://tracker.ceph.com/issues/42107
Signed-off-by: Shen Hang <harryshen18@gmail.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`

</details>

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -8916,12 +8923,26 @@ loff_t Client::_lseek(Fh *f, loff_t offset, int whence)
...
 
   case SEEK_END:
-    r = _getattr(in, CEPH_STAT_CAP_SIZE, f->actor_perms);
-    if (r < 0)
-      return r;
     pos = in->size + offset;
     break;
 
+  case SEEK_DATA:
+    if (offset >= in->size) {
```
Comment (jtlayton) [inline src/client/Client.cc]: offset is a signed value, so I think you need to check for it being negative as well. The kclient's llseek implementation does this for SEEK_DATA and SEEK_HOLE:
```
                if (offset < 0 || offset >= i_size) {
                        ret = -ENXIO;
                        goto out;
                }
```
I think you'll want to do the same here.

Comment (shenhang) [reply src/client/Client.cc]: > offset is a signed value, so I think you need to check for it being negative as well. The kclient's llseek implementation does this for SEEK_DATA and SEEK_HOLE:
> 
> ```
>                 if (offset < 0 || offset >= i_size) {
>                         ret = -ENXIO;
>                         goto out;
>                 }
> ```
> 
> I think you'll want to do the same here.

Thanks. Changed

**Comment by jtlayton** [comment]
Retest this please

**Comment by batrick** [comment]
retest this please

**Comment by shenhang** [comment]
> Please create a [tracker issue](https://tracker.ceph.com/) so that this issue can be tracked for backporting.
> 
> Then, please annotate the commit which fixes/resolves a Ceph tracker issue with:
> 
> `Fixes: http://tracker.ceph.com/issues/...`
> 
> This is essential when examining the history of the repository (this commit fixes what) and helps merge scripts identify issues that have been resolved by a merge. See [this article](https://help.github.com/en/articles/changing-a-commit-message) on GitHub on how to amend commits and update your pull request.

changed Thanks.

**Pull request merged**